core/src/main/java/com/rudderstack/android/sdk/core/FlushUtils.java

```diff
@@ -0,0 +1,244 @@
+                    if (networkResponse == Utils.NetworkResponses.SUCCESS) {
+                        // remove events from DB
+                        RudderLogger.logDebug(String.format("EventRepository: flush: Successfully sent batch %d/%d ", i, numberOfBatches));
+                        RudderLogger.logInfo(String.format(Locale.US, "EventRepository: flush: clearingEvents of batch %d from DB: %s", i, networkResponse));
+                        dbManager.clearEventsFromDB(batchMessageIds);
+                        messageIds.removeAll(batchMessageIds);
+                        messages.removeAll(batchMessages);
+                        lastBatchFailed = false;
+                        break;
+                    }
+                }
+                RudderLogger.logWarn(String.format("EventRepository: flush: Failed to send batch %d/%d retrying again, %d retries left", i, numberOfBatches, retries));
+            }
+            if (lastBatchFailed) {
+                RudderLogger.logWarn(String.format("EventRepository: flush: Failed to send batch %d/%d after 3 retries , dropping the remaining batches as well", i, numberOfBatches));
+                return false;
+            }
+        }
+        return true;
+
+
+    }
+
+    /*
+     * flush events payload to server and return response as String
+     * */
+    static Utils.NetworkResponses flushEventsToServer(String payload, String dataPlaneUrl, String authHeaderString, String anonymousIdHeaderString) {
+        try {
+            if (TextUtils.isEmpty(authHeaderString)) {
+                RudderLogger.logError("EventRepository: flushEventsToServer: WriteKey was not correct. Aborting flush to server");
+                return null;
+            }
+
+            // get endPointUrl form config object
+            String dataPlaneEndPoint = dataPlaneUrl + "v1/batch";
+            RudderLogger.logDebug("EventRepository: flushEventsToServer: dataPlaneEndPoint: " + dataPlaneEndPoint);
+
+            // create url object
+            URL url = new URL(dataPlaneEndPoint);
+            // get connection object
+            HttpURLConnection httpConnection = (HttpURLConnection) url.openConnection();
+            // set connection object to return output
+            httpConnection.setDoOutput(true);
+            //  set content type for network request
+            httpConnection.setRequestProperty("Content-Type", "application/json");
+            // set authorization header
+            httpConnection.setRequestProperty("Authorization", String.format(Locale.US, "Basic %s", authHeaderString));
+            // set anonymousId header for definitive routing
+            httpConnection.setRequestProperty("AnonymousId", anonymousIdHeaderString);
+            // set request method
+            httpConnection.setRequestMethod("POST");
+            // get output stream and write payload content
+            OutputStream os = httpConnection.getOutputStream();
+            synchronized (FLUSH_LOCK) {
```

Comment:
Can we add comments on what are we doing here with this lock?

Reply:
Sure.